tools/crm_mon.c

```diff
@@ -3000,21 +3006,44 @@ reduce_stonith_history(stonith_history_t *history)
     for (hp = history; hp; ) {
         for (np = new; np; np = np->next) {
             if ((hp->state == st_done) || (hp->state == st_failed)) {
+                enum redude_state reduce = reduce_none;
+
                 /* action not in progress */
                 if (safe_str_eq(hp->target, np->target) &&
-                    safe_str_eq(hp->action, np->action) &&
-                    (hp->state == np->state)) {
-                    if ((hp->state == st_done) ||
-                        safe_str_eq(hp->delegate, np->delegate)) {
-                        /* replace or purge */
-                        if (hp->completed < np->completed) {
-                            /* purge older hp */
-                            tmp = hp->next;
-                            hp->next = NULL;
-                            stonith_history_free(hp);
-                            hp = tmp;
-                            break;
+                    safe_str_eq(hp->action, np->action)) {
+                    if (hp->state == np->state) {
+                        if ((hp->state == st_done) ||
+                            safe_str_eq(hp->delegate, np->delegate)) {
+                            /* replace or purge */
+                            if (hp->completed < np->completed) {
+                                reduce = reduce_purge;
+                            }
+                            reduce = reduce_replace;
```

Comment:
Doesn't look as if that code makes much sense.
In the end you'll always get reduce_replace ...

Reply:
Sorry....I am making a mistake. I will review the correction.